Add has-matching-release-tag

jdbaldry opened the pull request:
Produce a boolean output that has the value "true" if and only if there is a release tag that
    matches the release branch.
    If the provided `ref_name` is a release tag reference, the output value is "true".
    If the provided `ref_name` is a release branch reference, the output value is "true" if
    and only if there is a matching release tag as defined by the `release_tag_regexp`.
    For release branches with a patch version, the capture groups for major, minor, and patch
    versions must match between the `release_branch_with_patch_regexp` and the
    `release_tag_regexp`.
    For release branches without a patch version, the capture groups for major and minor
    versions must match between the `release_branch_without_patch_regexp` and the
    `release_tag_regexp` and the release tag must have a `0` patch version.
    
Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry commented:
Hey all reviewers, I've added you all in case you are interested. Feel free to remove yourself if you would rather not get notified by all the conversation.

This is my proposal for gating publishing technical documentation to only release tags and releasing branches that have a corresponding release tag. This will then allow teams to work on release branches without worrying about publishing documentation until they are ready. As far as I'm aware, all repositories use tags for releases and release branches are long lived branches for backporting.

robbymilo commented:
The implementation on https://github.com/grafana/backend-enterprise/pull/3340 looks really good! I think this will be very useful across the org with it's flexibility.

jdbaldry commented on has-matching-release-tag/hasMatchingReleaseTag.ts:
```diff
@@ -0,0 +1,77 @@
+		},
+	}
+
+	await exec('git', ['tag'], options)
```
Ensure proper error handling of command not found or non zero exit code from command.

jdbaldry replied:
Also ensure that tags is not empty, if it is, encourage use of fetch-depth: 0

jdbaldry replied:
Added in 83b5e6f and 75e01ef

jdbaldry commented on has-matching-release-tag/action.yaml:
```diff
@@ -0,0 +1,36 @@
+    release_tag_regexp:
+        description: |
+            Regexp to match release tag references.
+            The regexp must have three capture groups, one for each of major, minor, and patch versions.
```
Add example for each

jdbaldry replied:
Added in 7f0e508

mdisibio commented on has-matching-release-tag/hasMatchingReleaseTag.ts:
```diff
@@ -0,0 +1,77 @@
+
+	let releaseTags = filterRefNames(refNames, releaseTagRegexp)
+
+	let branchMatches = refName.match(releaseBranchWithPatchRegexp)
```
As discussed, this use case is expected to be fairly uncommon.  I think we could make this optional by skipping if `releaseBranchWithPatchRegexp` is an empty string?   If so, it might be sensible to make that the default input value as well, and workflows wouldn't have to define it.

jdbaldry replied:
Made optional in a92ad8b

mdisibio commented on has-matching-release-tag/hasMatchingReleaseTag.ts:
```diff
@@ -0,0 +1,77 @@
+		}
+	}
+
+	branchMatches = refName.match(releaseBranchWithoutPatchRegexp)
```
Thinking about if this check could be moved ahead of the previous one since it is more common, from a readability point of view.

jdbaldry replied:
Reordered in a92ad8b

mdisibio commented:
👍 Yes this is excellent.  I think the main thing I'd like is to make the `release_branch_with_patch_regexp` input and use case optional, since we don't need it in Tempo or GET.

qkombur commented:
I have nothing to add other then typescript looks really cool.

jdbaldry commented:
Thanks both for taking a look. I've addressed some of the review feedback and also taking the opportunity to replace the asynchronous exec with child_process.execFileSync since I don't need async execution.

For an example of usage, please take a look at the PR in grafana/backend-enterprise https://github.com/grafana/backend-enterprise/pull/3340

mdisibio commented on has-matching-release-tag/hasMatchingReleaseTag.ts:
```diff
@@ -0,0 +1,84 @@
+				tagMatches[2] == branchMatches[2] &&
+				tagMatches[3] == '0'
+			) {
+				return 'true'
```
For troubleshooting, could we log something for the 3 cases when they return true?  For this case something like: `"Found matching release tag: " + releaseTags[i] + " in branch " + refName`

mdisibio commented:
I was able to test this out for Tempo/GET use cases and it is looking good.  

(1) Workflow is [here](https://github.com/mdisibio/tempo/blob/release-0.999/.github/workflows/test.yml#L33)

(2) Test commit to release-0.998 branch with no .0 tag.  Correctly logs and skips the next steps:
<img width="620" alt="image" src="https://user-images.githubusercontent.com/13524475/166814153-41f004d6-90b5-4193-a2de-1139f346cfec.png">

(3) Test commit to release-0.999 where 0.999.0 tag exists.  Logs and runs the next steps:
<img width="670" alt="image" src="https://user-images.githubusercontent.com/13524475/166814327-66c681bf-df97-46ba-98bf-07245110e71d.png">

jdbaldry commented:
Thanks Marty! I will get this merged today and start deploying this to repositories as part of https://github.com/grafana/technical-documentation/issues/297
